### ISSUE TITLE
Apps position is not the same showed in the iTunes Store

### ISSUE THREAD

**Issue by icaroponce** [facundoolano/app-store-scraper#30]
Hello there, 

I'm using the module to search for some apps, but the result returned is not the same that I get on the iTunes app (Yes, I had set the country correctly) which works perfectly in the `google-play-scraper` module. As I'm using it for ASO purposes, it's critical to know the real apps positions in the iTunes. (For example, I want to know if the app which i'm optimizing in ranking better, if it's in the first position for a given word...)

Talking to a friend of mine who works with ASO daily, he told me that the iTunes API is not reliable when it comes to the apps positioning (maybe for a delay to update, idk, but it isn't the same list) and another solution would be necessary.

Had you guys have the same problems or notice something like this? Or am I the only one?

Actually, if it's really a problem, I'm considering a solution: I've found out using a Sniffer (like Wireshark), to monitor the network when I'm using the iTunes app, and the app do requests to another url, setting different headers to fetch the content to be presented on it...

Please, let me know what you guys think about it.

**Comment by facundoolano** [comment]
I haven't really compared both sets of search results, but if you've found differences it would definitely make sense trying to look into the actual APIs the iTunes app is using. I'd really be interested in seeing those in case you can take the time to dig around using a sniffer.

**Comment by icaroponce** [comment]
Sure,

here an example of url the app hits:

[https://search.itunes.apple.com/WebObjects/MZSearchHints.woa/wa/hints?q=panda](https://search.itunes.apple.com/WebObjects/MZSearchHints.woa/wa/hints?q=panda)

Just sending a get request to it, you'll get a XML file.

You can set params to it through the header:

```{"X-Apple-Tz":"-7200"}
{"Accept-Language":"pt-br"}
{"X-Apple-Store-Front":"143503-15,32"}
{"X-Apple-I-MD-RINFO":"17106176"}
{"X-Apple-I-MD-M":"5qf7Ex9VvAxuH98/nDvbIHqW76fYyHTpVEhcWC2tBT8FKc7zqErTS9/kmJlqWXGSfPgcdl831PNkbddc"}
{"X-Apple-I-MD":"AAAABQAAABCDXFWHMZGdT6C1K2IFjB00AAAAAg=="}
```

These last  two tokens represent language and country. 

I'm quite busy this week, but if you can't check it yourself during the week, I could dig it deeper on the weekend. 

But, all we need is to turn on the sniffer and watch the requests behavior to understand all the params sent on header. 

I know it's a more verbose way to do it, but if really there's difference in the api, this way seems to be the most accurate way to retrieve data for some methods, like search for example.

**Comment by facundoolano** [comment]
Mmm the hints endpoint you show is for search hints, not search itself. That endpoint is already being used for the `suggest` function ([see this](https://github.com/facundoolano/app-store-scraper/blob/master/lib/suggest.js#L6)). I imagine the search itself uses a different endpoint.

**Comment by icaroponce** [comment]
Ops, I'm sorry. Got the wrong endpoint.

Check it:

```
curl -X GET \
  'https://itunes.apple.com/WebObjects/MZStore.woa/wa/search?submit=edit&mt=1&term=pandas' \
  -H 'accept-language: en-us' \
  -H 'x-apple-i-md: AAAABQAAABASYMVHFwrAt3wPSUwbTCqGAAAAAg==' \
  -H 'x-apple-i-md-m: hxpW6pCpiMIBZ+ch0fA6Hjke7zRLyqe5KdegwA9TzU2aDGPWCL1CX85fi+s4dFX4yDC58Owm+DJImUbI' \
  -H 'x-apple-store-front: 143441-1,32 ab:IMP0' \
  -H 'x-apple-tz: -7200'
```
It'll return a HTML file. A bit confusing.

But, search a for a <script> tag that has `its.serverData`. You'll notice that it's a JSON. Get it. 

Access its property `pageData.searchPageData.bubbles`. It'll bring an array that aggregates data by its kind. The index 3 `bubbles[3]`, whose name proprety is 'software', has a property results which contains 116 results. 

Each item of array has an id property. its the app id.

The most important here: these results array in are "ordered correctly", which means that it brings the same apps order that I get when a search for something in my iPhone or iTunes APP, which is not true when using iTunes API. 

Hope it'll help.

**Comment by anater** [comment]
Has this issue been worked on?

**Comment by facundoolano** [comment]
Nope, no updates.

**Comment by cloforce** [comment]
@icaroponce Does your App Store Scraper use the endpoint you talk about here? So far it is in fact the only endpoint I can find that returns the same ordered results as the app.

**Comment by icaroponce** [comment]
Hey @cloforce,

Yes, it does. 

We had to develop a small internal lib to use the endpoint pointed above, seeing that the real current position for a given app is very important to our platform we had to do that.

**Comment by cloforce** [comment]
@icaroponce Awesome! Would you guys be open to sharing that internal library? Can email me privately about it.

**Comment by anater** [comment]
To be clear, the endpoint you’re talking about is in use in this library: https://github.com/facundoolano/app-store-scraper

I’m still looking for the best way to get App Store search results that reflect the ones in iOS 11

**Comment by anater** [comment]
That was a question, btw 😅

**Comment by icaroponce** [comment]
No @anater,  It's not. Otherwise this issue would be already closed. 

If you follow my instructions you will get the list that reflects the App Store. Pay attention to the headers set only, because you've to set it properly to a specific country and language.

**Comment by icaroponce** [comment]
@cloforce unfortunately I'm not working at that company anymore. That time we gave up opening it because we wanted update this lib or even add another endpoint here instead of creating a new lib to do almost the same thing. 

I still can check with the guys if they would be up to share the lib though without problems.

**Comment by cloforce** [comment]
@icaroponce Thanks for the reply. Any update from the guys? If not I think I'm going to begin building it out myself. Let me know, appreciate it.

**Comment by icaroponce** [comment]
Hey @cloforce,

UnfortunateIy I haven't got a reply yet.

You better start your own, I'll let you know in the case something changes.

**Comment by facundoolano** [comment]
@icaroponce can you refresh my memory a bit here? On May 4, 2017 you described a way to get the properly ordered app list. Does that list includes all the app data or it only servers to get the ordered ids?

**Comment by icaroponce** [comment]
Yes, this url provides a correctly ordered `appId`s list. Then we used another endpoint, ( I guess that was the`/lookup`) to fetch app details. This endpoint allows us to send more than 100 `appId`s at once, then we basically could fetch everything in 3 requests (1 for all the ordered ids, 2 in order to fetch the details ). I also need to refresh my memory about some details, but as soon as a I have a free time I'll investigate it again and comment down here everything more elaborated and how to reproduce the complete flow.

**Comment by facundoolano** [comment]
I'm looking into this.

@icaroponce the current search endpoint allows to filter by device (ipad, mac or all), do you know if such filtering is available with the https://itunes.apple.com/WebObjects/MZStore.woa/wa/search endpoint?

**Comment by facundoolano** [comment]
This one seems to give the json directly

``` 
curl -X GET \
 'https://itunes.apple.com/WebObjects/MZStore.woa/wa/search?clientApplication=Software&media=software&term=panda' \
 -H 'accept-language: en-us' \
 -H 'X-Apple-Store-Front: 143465-19,24 t:native'
```

**Comment by DataGreed** [comment]
@facundoolano did you find a way to filter by device (ipad/iphone)?

**Comment by master4apple** [comment]
How can i unlock any iPhone Device i have mobile shop please help

**Comment by ejhg** [comment]
The integer after the comma in the X-Apple-Store-Front header allows you to filer by iPad or iPhone.  That value seems to change over time based on examples I've found across the web.  Today, "143441,29" returns results for iPhone (aka "software") and "143441,30" returns results for iPad aka "iPadSoftware".

(I still haven't figured out the value for "macSoftware" for macOS.)

**Comment by Nano112** [comment]
> The integer after the comma in the X-Apple-Store-Front header allows you to filer by iPad or iPhone. That value seems to change over time based on examples I've found across the web. Today, "143441,29" returns results for iPhone (aka "software") and "143441,30" returns results for iPad aka "iPadSoftware".
> 
> (I still haven't figured out the value for "macSoftware" for macOS.)
> 
> Also, you don't need the `accept-language: en-us` header nor the `t:native` as far as I can tell.
> 
> This works well
> 
> ```
> curl -H 'x-apple-store-front: 143441,29' 'https://search.itunes.apple.com/WebObjects/MZStore.woa/wa/search?clientApplication=Software&term=panda'
> ```
> 
> Finally, note that the search endpoint hostname is now `search.itunes.apple.com`, not `itunes.apple.com`.

I found this if ever it is helpful to anyone:
https://stackoverflow.com/questions/47406922/what-does-x-apple-store-front-means-in-apple-http-headers

**Comment by andresbeef** [comment]
Hi all,

Anyone found a solution to get the exact same results that you get from the iphone?
